Add c?r?(begin|end) functions for ezArrayMap…

JohannStudanski opened the pull request:
… to make it work in range based for loops (etc.), incl. test.

JohannStudanski committed:
Add c?r?(begin|end) functions for ezArrayMap to make it work in range based for loops (etc.), incl. test.

jankrassnigg commented on Code/Engine/Foundation/Containers/Implementation/ArrayMap_inl.h:
```diff
@@ -176,6 +176,18 @@ VALUE& ezArrayMapBase<KEY, VALUE>::GetValue(ezUInt32 index)
 }
 
+template <typename KEY, typename VALUE>
+EZ_ALWAYS_INLINE ezDynamicArray<typename ezArrayMapBase<KEY, VALUE>::Pair>& ezArrayMapBase<KEY, VALUE>::GetData()
```
This would allow people to modify the keys and thus destroy the sorting. I would reset m_bSorted to false here, to guarantee a sorting step before the next lookup. This is not necessary for the const version.

jankrassnigg replied:
You could add a unit test for modifying the keys, and checking that the still lookup works afterwards, if you like, though I don't see that as super important.

JohannStudanski replied:
Good point! I was torn on exposing the method publicly or not doing it and making the begin and end methods friends, Marc favored the public method, so I did that. Should've then added at least the m_bSorted update for the non-const method and a test tho, totally right, I'll do that now.

jankrassnigg commented on Code/UnitTests/FoundationTest/Containers/ArrayMapTest.cpp:
```diff
@@ -7,6 +7,90 @@
+    for (ezUInt32 i = 0; i < 1000; ++i)
+      m[i] = i + 1;
+
+    //{ // No finding by key (or value) using std::find without a custom comparator functor (same as in MapTest)
+    //  auto itfound = std::find(begin(m), end(m), 500);
+    //  EZ_TEST_INT(itfound->key, 499);
+    //}
+
+    //{ // No finding by key (or value) using std::find without a custom comparator functor (same as in MapTest)
+    //  auto itfound = std::find(rbegin(m), rend(m), 500);
+    //  EZ_TEST_INT(itfound->key, 499);
+    //}
```
Is this supposed to get activated?

JohannStudanski replied:
I don't know! I saw the same code commented out in the MapTest.cpp (Lines 16 & 20), so I added them here as well. I thought they were left commented out there so that e.g. if there's a push to add the comparators it's a little more obvious where they should be tested or something. In any case, I just wanted to mirror the code there.

If you'd rather I just leave it out I can of course do that, too.

jankrassnigg replied:
Yeah, remove it. Or try if it would compile now. The comment in the map test is probably very, very old. Maybe it would even compile now? It might work since we added begin/end for ezMap.

JohannStudanski replied:
No it doesn't (it can't, not even in the std:::maps, and I tried): std::find runs an equality comparison between the value you pass and each element in the iterable, but since the iterable elements are pairs in a map, you can only either a) pass a kv-Pair (it'll find that of course), or b) pass in a key or value AND a functor that does the comparison correctly.

It would of course be possible to add the functor—and/or a Find method that searches for values instead of keys... ¯\_(ツ)_/¯

jankrassnigg replied:
Then just get rid of it.

jankrassnigg reviewed (approved):
Great addition ! I was aware of this shortcoming but had not time to do this myself, so this is really welcome :-)

jankrassnigg commented:
/AzurePipelines run

JohannStudanski committed:
Fix for ezArrayMap's new GetData method to mark the map as unsorted after passing out the non const data reference.
Additional tests to check if map indexing and iteration still work after modifying the keys via the non-const data reference.
